File: shared/Sync/MatchSyncCoordinator.swift
import Foundation
import WatchConnectivity

/// Bidirectional sync of active match + archive between Watch and iPhone.
/// Watch is authoritative during live scoring; iPhone applies remote snapshots.
/// History deletion runs the other way: the iPhone owns it, and the Watch prunes
/// its own archive to match.
@MainActor
public final class MatchSyncCoordinator: NSObject, ObservableObject {
    public static let activeKey = "activeMatchJSON"
    public static let archiveKey = "archiveJSON"
    public static let deletedKey = "deletedMatchIDsJSON"

    private let service: MatchService
    private let session: WCSession?
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder
    private let isWatch: Bool

    public init(service: MatchService, isWatch: Bool) {
        self.service = service
        self.isWatch = isWatch
        self.encoder = JSONEncoder()
        self.encoder.dateEncodingStrategy = .iso8601
        self.decoder = JSONDecoder()
        self.decoder.dateDecodingStrategy = .iso8601

        if WCSession.isSupported() {
            self.session = WCSession.default
        } else {
            self.session = nil
        }

        super.init()

        if let session {
            session.delegate = self
            session.activate()
        }

        service.onSyncNeeded { [weak self] active, archive, deleted in
            self?.push(active: active, archive: archive, deleted: deleted)
        }
    }

    public func push(active: MatchState?, archive: [MatchState], deleted: Set<UUID>) {
        guard let session, session.activationState == .activated else { return }

        var context: [String: Any] = [:]
        do {
            if let active {
                context[Self.activeKey] = try encoder.encode(active)
            } else {
                context[Self.activeKey] = Data()
            }
            context[Self.archiveKey] = try encoder.encode(archive)
            // Sent in full rather than as a delta, so a Watch that was unreachable at
            // deletion time still catches up on its next context update.
            context[Self.deletedKey] = try encoder.encode(Array(deleted))
            try session.updateApplicationContext(context)

            #if os(watchOS)
            if session.isReachable {
                session.sendMessage(context, replyHandler: nil, errorHandler: nil)
            }
            #else
            if session.isReachable {
                session.sendMessage(context, replyHandler: nil, errorHandler: nil)
            }
            #endif
        } catch {
            // Non-fatal: local scoring continues offline.
        }
    }

    private func apply(context: [String: Any]) {
        let deleted: Set<UUID>
        if let data = context[Self.deletedKey] as? Data, !data.isEmpty {
            deleted = Set((try? decoder.decode([UUID].self, from: data)) ?? [])
        } else {
            deleted = []
        }

        // Watch remains authoritative for live scoring, so it takes deletions from the
        // phone and ignores everything else in the payload.
        guard !isWatch else {
            service.applyRemoteDeletions(deleted)
            return
        }

        let active: MatchState?
        if let data = context[Self.activeKey] as? Data, !data.isEmpty {
            active = try? decoder.decode(MatchState.self, from: data)
        } else {
            active = nil
        }

        let archive: [MatchState]
        if let data = context[Self.archiveKey] as? Data, !data.isEmpty {
            archive = (try? decoder.decode([MatchState].self, from: data)) ?? []
        } else {
            archive = []
        }

        service.applyRemoteSnapshot(active: active, archive: archive)
    }
}

extension MatchSyncCoordinator: WCSessionDelegate {
    nonisolated public func session(
        _ session: WCSession,
        activationDidCompleteWith activationState: WCSessionActivationState,
        error: Error?
    ) {
        Task { @MainActor in
            if activationState == .activated {
                self.push(
                    active: self.service.activeMatch,
                    archive: self.service.archivedMatches,
                    deleted: self.service.deletedMatchIDs
                )
            }
        }
    }

    #if os(iOS)
    nonisolated public func sessionDidBecomeInactive(_ session: WCSession) {}
    nonisolated public func sessionDidDeactivate(_ session: WCSession) {
        session.activate()
    }
    #endif

    nonisolated public func session(
        _ session: WCSession,
        didReceiveApplicationContext applicationContext: [String: Any]
    ) {
        Task { @MainActor in
            self.apply(context: applicationContext)
        }
    }

    nonisolated public func session(
        _ session: WCSession,
        didReceiveMessage message: [String: Any]
    ) {
        Task { @MainActor in
            self.apply(context: message)
        }
    }
}

File: android/sync/src/main/kotlin/com/wristrally/sync/MatchSyncCoordinator.kt
package com.wristrally.sync

import com.wristrally.domain.MatchService
import com.wristrally.domain.MatchState
import java.util.UUID

/**
 * Bidirectional sync of active match + archive between Wear OS and the phone.
 * Watch is authoritative during live scoring; the phone applies remote snapshots.
 * History deletion runs the other way: the phone owns it, and the watch prunes
 * its own archive to match.
 */
class MatchSyncCoordinator(
    private val service: MatchService,
    private val isWatch: Boolean,
    private val transport: SyncTransport,
) {
    init {
        transport.setListener { path, payload, fromLocalNode ->
            apply(path, payload, fromLocalNode)
        }
        service.onSyncNeeded { active, archive, deleted ->
            push(active, archive, deleted)
        }
        push(service.activeMatch, service.archivedMatches, service.deletedMatchIDs)
    }

    fun apply(path: String, payload: ByteArray, fromLocalNode: Boolean) {
        if (fromLocalNode) return
        val decoded = runCatching { SyncCodec.decode(payload) }.getOrNull() ?: return
        if (isWatch) {
            if (path == SyncPaths.DELETED) {
                service.applyRemoteDeletions(decoded.deleted)
            }
        } else if (path == SyncPaths.SNAPSHOT) {
            service.applyRemoteSnapshot(decoded.active, decoded.archive)
        }
    }

    fun push(active: MatchState?, archive: List<MatchState>, deleted: Set<UUID>) {
        val payload = SyncCodec.encode(SyncPayload(active, archive, deleted))
        val path = if (isWatch) SyncPaths.SNAPSHOT else SyncPaths.DELETED
        transport.publish(path, payload)
        transport.sendMessage(path, payload)
    }
}
